src/elona/draw.cpp

```diff
@@ -706,170 +716,168 @@ void create_pcpic(int cc, bool with_equipments)
         {
-            chipm(0, cnt) = 10;
-            chipm(1, cnt) = 9;
+            shadow = 40;
         }
-        for (int cnt = 396; cnt < 825; ++cnt)
+        if (game_data.weather == 4 && shadow < 65)
         {
-            chipm(7, cnt) = 5;
+            shadow = 65;
         }
-        for (int cnt = 264; cnt < 297; ++cnt)
+        if (game_data.current_map == mdata_t::MapId::noyel &&
+            (game_data.date.hour >= 17 || game_data.date.hour < 7))
         {
-            chipm(7, cnt) = 4;
+            shadow += 35;
         }
-        chipm(5, 135) = 8;
-        chipm(5, 137) = 16;
-        chipm(5, 140) = 6;
-        chipm(5, 145) = 16;
-        chipm(5, 149) = 16;
     }
-    if (map_data.atlas_number == 1)
+
+    return shadow;
+}
+
+
+
+void draw_prepare_map_chips()
+{
+    map_tileset(map_data.tileset);
+
+    int shadow = _get_map_chip_shadow();
+    snail::Color color{(uint8_t)(255 - shadow)};
+
+    bool changed = false;
+
+    gmode(0);
+    for (const auto& buffer :
+         loader.get_buffers_of_type(PicLoader::PageType::map_chip))
     {
-        for (int cnt = 396; cnt < 825; ++cnt)
-        {
-            chipm(7, cnt) = 5;
-        }
-        chipm(5, 233) = 56;
-        chipm(6, 233) = 48;
-        chipm(7, 594) = 4;
-        chipm(7, 628) = 4;
-        chipm(7, 637) = 4;
-        chipm(7, 641) = 4;
-        chipm(7, 733) = 4;
-        for (int cnt = 45; cnt < 61; ++cnt)
-        {
-            chipm(0, cnt) = 4;
-        }
-        chipm(0, 82) = 4;
-        chipm(0, 83) = 4;
-        chipm(0, 84) = 4;
-        for (int cnt = 462; cnt < 528; ++cnt)
-        {
-            chipm(2, cnt) = 1;
-            chipm(2, cnt - 66) = 1;
-        }
-        for (int cnt = 462; cnt < 495; ++cnt)
-        {
-            chipm(2, cnt) = 2;
-            chipm(2, cnt - 66) = 2;
-        }
-        chipm(0, 29) = 1;
-        chipm(0, 30) = 2;
-        chipm(0, 31) = 2;
-        chipm(0, 464) = 6;
-        chipm(3, 550) = 2;
-        chipm(2, 550) = 1;
-        chipm(0, 165) = 3;
-        chipm(3, 165) = 3;
-        chipm(0, 168) = 3;
-        chipm(3, 168) = 3;
-        chipm(0, 171) = 3;
-        chipm(1, 171) = 5;
-        chipm(3, 171) = 3;
-        chipm(0, 594) = 3;
-        chipm(3, 594) = 3;
+        changed |= tinted_buffers.tint(buffer, color);
     }
-    if (map_data.atlas_number == 2)
+
+    if (changed)
     {
-        for (int cnt = 0; cnt < 11; ++cnt)
+        for (const auto& buffer :
+             loader.get_buffers_of_type(PicLoader::PageType::map_feat))
         {
-            int cnt2 = cnt;
-            for (int cnt = 0; cnt < 13; ++cnt)
-            {
-                chipm(0, cnt2 * 33 + cnt + 20) = 4;
-            }
-        }
-        for (int cnt = 33; cnt < 66; ++cnt)
-        {
-            chipm(0, cnt) = 4;
+            gmode(0);
+            tinted_buffers.tint(buffer, color);
+            gmode(2, 30);
+            tinted_buffers.tint(buffer, color, true);
```

Comment:
In the second call of `TintedBuffers::tint()`, should argument of `color` be `snail::Color{255}`?

```cpp
// Part of map_prepare_tileset_atlas()
    set_color_mod(255 - shadow, 255 - shadow, 255 - shadow, 6);     // set color modifier
    gcopy(6, 0, 0, 33 * inf_tiles, 25 * inf_tiles, 0, 0);      // draw all map chips including feat chips with the tint.
    set_color_mod(255, 255, 255, 6);        // clear the color modifier
    gmode(2, 30);        // set blend mode and opacity.
    if (map_data.atlas_number == 0)
    {
        gcopy(6, 0, 192, 1360, 48, 0, 192);       // draw feat chips
    }
    ...
```

So in original code, feat chips were drawn twice, drawn with the same tint as other map chips, and then drawn over with alpha 30 and without tint (in original color).